p2p/simulations/network_test.go

```diff
@@ -19,13 +19,119 @@ package simulations
 import (
 	"context"
 	"fmt"
+	"os"
 	"testing"
 	"time"
 
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
 	"github.com/ethereum/go-ethereum/p2p/enode"
 	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	"github.com/ethereum/go-ethereum/rpc"
 )
 
+func init() {
+	log.Root().SetHandler(log.LvlFilterHandler(4, log.StreamHandler(os.Stderr, log.TerminalFormat(true))))
+}
+
+func TestSnapshotExplicit(t *testing.T) {
+	// create simulation network with 20 testService nodes
+	adapter := adapters.NewSimAdapter(adapters.Services{
+		"dummy":  newDummyService,
+		"dummy2": newDummy2Service,
+	})
+	network := NewNetwork(adapter, &NetworkConfig{
+		DefaultService: "dummy",
+	})
+	defer network.Shutdown()
+	nodeCount := 3
+	ids := make([]enode.ID, nodeCount)
+	for i := 0; i < nodeCount; i++ {
+		conf := adapters.RandomNodeConfig()
+		conf.Services = []string{"dummy", "dummy2"}
+		node, err := network.NewNodeWithConfig(conf)
+		if err != nil {
+			t.Fatalf("error creating node: %s", err)
+		}
+		if err := network.Start(node.ID()); err != nil {
+			t.Fatalf("error starting node: %s", err)
+		}
+		ids[i] = node.ID()
+	}
+	var eventsDone = make(chan struct{})
+	count := 2
+	events := make(chan *Event)
+	sub := network.Events().Subscribe(events)
+	go func() {
+		for event := range events {
+			if event.Type == EventTypeConn && !event.Control {
+				count--
+				if count == 0 {
+					eventsDone <- struct{}{}
+					return
+				}
+			}
+		}
+	}()
+	err := network.Connect(ids[0], ids[1])
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = network.Connect(ids[0], ids[2])
+	if err != nil {
+		t.Fatal(err)
+	}
+	<-eventsDone
+	go func() {
+		defer sub.Unsubscribe()
+		for event := range events {
+			if event.Type == EventTypeConn && !event.Control {
+				eventsDone <- struct{}{}
+				return
+			}
+		}
+	}()
+	err = network.Disconnect(ids[0], ids[2])
+	if err != nil {
+		t.Fatal(err)
+	}
+	<-eventsDone
+
+	snap, err := network.Snapshot()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(snap.Conns) > 1 {
+		t.Fatalf("expected one connect object")
+	}
+	for _, svc := range snap.Nodes[0].Node.Config.Services {
+		if svc != "dummy" && svc != "dummy2" {
+			t.Fatalf("unexpected service %s", svc)
+		}
+	}
+
+	snap, err = network.SnapshotWithServices([]string{"bzz"}, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, svc := range snap.Nodes[0].Node.Config.Services {
+		if svc != "dummy" && svc != "dummy2" && svc != "bzz" {
+			t.Fatalf("unexpected service %s", svc)
+		}
+	}
+
+	snap, err = network.SnapshotWithServices([]string{"bzz"}, []string{"test"})
```

Comment:
I guess it is best to put `dummy2` in the services to be removed not the non-existing `test` and then check it it is removed on line 129